contracts/SponsorFaucet.sol

```diff
@@ -64,38 +109,180 @@ contract SponsorFaucet is Ownable, Pausable, ReentrancyGuard {
+        uint256 upperBound
+    ) public {
+        _onlyWhitelistAdmin();
+        //rule by internal contract
+        require(upperBound.mul(1000) <= gasBound, "upperBound too high");
+        general_gas_total_limit = gasTotalLimit;
+        general_collateral_total_limit = collateralTotalLimit;
+        general_gas_bound = gasBound;
+        general_collateral_bound = collateralBound;
+        general_upper_bound = upperBound;
+    }
+
+    //set bounds for special contracts
+    function setSpecialBounds(
+        address addr,
+        uint256 gasTotalLimit,
+        uint256 collateralTotalLimit,
+        uint256 gasBound,
+        uint256 collateralBound,
+        uint256 upperBound
+    ) public {
+        _onlyWhitelistAdmin();
+        //rule by internal contract
+        require(upperBound.mul(1000) <= gasBound, "upperBound too high");
+        special_dapps[addr].gas_total_limit = gasTotalLimit;
+        special_dapps[addr].collateral_total_limit = collateralTotalLimit;
+        special_dapps[addr].gas_bound = gasBound;
+        special_dapps[addr].collateral_bound = collateralBound;
+        special_dapps[addr].upper_bound = upperBound;
+    }
+
+
+    /* ===== Public utility functions ===== */
+    function addGeneralContracts(address[] memory addAddr) public {
+        _onlyWhitelistAdmin();
+        for(uint256 i = 0; i < addAddr.length; i++) {
+            general_contracts.push(addAddr[i]);
+        }
+    }
+
+    function addSpecialContracts(address[] memory addAddr) public {
+        _onlyWhitelistAdmin();
+        for(uint256 i = 0; i < addAddr.length; i++) {
+            special_contracts.push(addAddr[i]);
+        }
+    }
+
+    function specialContractsCount() public view returns (uint256) {
+        return special_contracts.length;
+    }
+
+    function generalContractsCount() public view returns (uint256) {
+        return general_contracts.length;
+    }
+
+    function isGeneralContract(address dapp) public view returns (bool) {
+        for(uint256 i = 0; i < general_contracts.length; i++) {
+            if(general_contracts[i] == dapp) return true;
+        }
+        return false;
+    }
+
+    function isSpecialContract(address dapp) public view returns (bool) {
+        for(uint256 i = 0; i < special_contracts.length; i++) {
+            if(special_contracts[i] == dapp) return true;
+        }
+        return false;
+    }
+
+    function _onlyWhitelistAdmin() internal view {
+        require(
+            isWhitelistAdmin(msg.sender),
+            "caller does not have the WhitelistAdmin role"
+        );
+    }
+
     function _isAppliableForGas(address dapp) internal returns (bool) {
+        if(!dapp.isContract()) return false;
         uint256 gas_balance = internal_sponsor.getSponsoredBalanceForGas(dapp);
         address current_sponsor = internal_sponsor.getSponsorForGas(dapp);
+        uint256 contract_gas_bound;
+        uint256 contract_gas_total_limit;
+        if (isGeneralContract(dapp)) {
```

Comment:
custom sponsorship first, then large, then small.

Reply:
default is small, then custom, then large